Adds in memory cache for RPC query results

reinhapa opened the pull request:
Enables in-memory query result values less than 4 MB instead of writing them to a temporary file.

By using the in-memory-buffer no potential virus scanner is trigger either on the server as on the client side. This almost doubled the performance on our productive systems, when using this change.

adamretter commented on exist-core/src/main/java/org/exist/xmlrpc/RpcConnection.java:
```diff
@@ -732,31 +722,34 @@ public Map<String, Object> getDocumentData(final String docName, final Map<Strin
         final byte[] chunk = new byte[len];
-        try (final InputStream is = Files.newInputStream(file)) {
-            is.skip(offset);
+        try (final InputStream is = file.newInputStream()) {
```
Should we buffer the I/O here perhaps?

reinhapa replied:
@adamretter the buffering is done in the ContentFile itself if not based on memory buffer already...

adamretter commented on exist-core/src/main/java/org/exist/xmlrpc/RpcConnection.java:
```diff
@@ -1155,21 +1148,21 @@ private Map<String, List> listDocumentPermissions(final XmldbURI collUri) throws
 
     @Override
-    public Map<XmldbURI, List> listCollectionPermissions(final String name)
+    public Map<String, List<Object>> listCollectionPermissions(final String name)
```
I am interested to know the thinking behind why this should be changed from `XmldbURI` to `String` please?

adamretter replied:
Presumably this was a new test that you added that was previously missing?

reinhapa replied:
Correct, among a lot of others 😎

adamretter reviewed (commented):
This sounds interesting and like you are seeing a good performance improvement. If I understand correctly, when the result of an RPC call is less than 4 MB in size, then instead of writing to a file, you keep it in memory for a finite period of time. Do I have that correct?

If that is the case, then in general that sounds fine, but I have some questions which I can't spot easily in your code (perhaps you can point me to specific parts of your PR):

1. Presumably the data coming from the result of RPC call is in one of two forms - either (a) an InputStream that is read directly from the network socket, or (b) a buffer that is already in-memory?

If its scenario (a), how do you know the size without reading all of the data? or do you just read up to 4 MB, and if it is 4MB or more, you flush the already read 4 MB to a file and then keep appending to the file until the stream is completely read?

If its scenario (b), are you avoiding making a second copy of the same data in-memory?

2. How are you bounding the total size of the cache? i.e. How do I make sure I don't have lots of 4 MB chunks of memory in-flight that ultimately add up to a large number and cause memory starvation?

Kind regards.

reinhapa commented:
> This sounds interesting and like you are seeing a good performance improvement. If I understand correctly, when the result of an RPC call is less than 4 MB in size, then instead of writing to a file, you keep it in memory for a finite period of time. Do I have that correct?

Correct.

> If that is the case, then in general that sounds fine, but I have some questions which I can't spot easily in your code (perhaps you can point me to specific parts of your PR):
> 
> 1. Presumably the data coming from the result of RPC call is in one of two forms - either (a) an InputStream that is read directly from the network socket, or (b) a buffer that is already in-memory?
> 
> If its scenario (a), how do you know the size without reading all of the data? or do you just read up to 4 MB, and if it is 4MB or more, you flush the already read 4 MB to a file and then keep appending to the file until the stream is completely read?

Absolutely correct. This is done within the existing `VirtualTempPath` implementation introduced earlier on the client side within `AbstractRemoteResource`, `RemoteResourceSet`, `RemoteXMLResource` and `InternalXMLSerializer`

> If its scenario (b), are you avoiding making a second copy of the same data in-memory?
> 
> 2. How are you bounding the total size of the cache? i.e. How do I make sure I don't have lots of 4 MB chunks of memory in-flight that ultimately add up to a large number and cause memory starvation?

At the moment, there is no such handling for permitting such logic. The current `MemoryContentsImpl` class holding the actual data stores those in 4k blocks within a two dimensional array instead of using one big 4MB  byte array.

adamretter commented:
> Absolutely correct. This is done within the existing VirtualTempPath implementation introduced earlier on the client side

I think you are referring specifically to the `OverflowToDiskStream` class, is that correct?

> At the moment, there is no such handling for permitting such logic that may be added later...

Just my opinion, but if we are going to Cache things in memory we should have an upper bound on the total size of that cache otherwise it will lead to problems sooner or later.

> The current `MemoryContentsImpl` class holding the actual data stores those in 4k blocks within a two dimensional array instead of using one big 4MB byte array.

An interesting design! One concern, with that sizing, that is potentially a lot of small allocations, and as these buffers (as I think I understand it) are not from a pool, it will also mean a lot of GC.

eXist-db does have bounded caches already based on Caffeine, and also has some Pooling based on custom mechanisms such as the simplistic `ByteArrayPool` or the more complete Apache Commons Pool.

reinhapa commented:
> I think you are referring specifically to the `OverflowToDiskStream` class, is that correct?

Yes, that's exactly the implementation...

> Just my opinion, but if we are going to Cache things in memory we should have an upper bound on the total size of that cache otherwise it will lead to problems sooner or later.

That is correct, I do not know how much the RPC option is actually used and more enhancements in that diection are shure an option.

> An interesting design! One concern, with that sizing, that is potentially a lot of small allocations, and as these buffers (as I think I understand it) are not from a pool, it will also mean a lot of GC.

In case of pooling the already used memory instances would be beneficial in that regard I guess...

> eXist-db does have bounded caches already based on Caffeine, and also has some Pooling based on custom mechanisms such as the simplistic `ByteArrayPool` or the more complete Apache Commons Pool.

I can look into this a bit more in the next days. The question is where to configure and monitor all those parts... Do monitor exising caches that you mentioned?

adamretter commented:
> I do not know how much the RPC option is actually used

Hmm... My awareness is that it is used by Oxygen XML Editor, The Java Admin Client, and xst.
However they are all pretty much single-user uses. So not bounding the cache there won't make much difference.

But I guess like you, others are using it for integration - if I recall correctly, I think the Python libraries for eXist-db use the XML-RPC interface (as of course does the XML:DB interface that you wrote/updated).

reinhapa commented:
> For monitoring - some of it is exposed via JMX (particularly Caffeine and some of the Apache Commons pools). For sizing and configuration of this stuff, I think its normally done from `conf.xml` at startup.

@adamretter do you got any suggestions, where there is the best place to configure that to be created `ContentFilePool` within the `conf.xml`?

adamretter commented:
> do you got any suggestions, where there is the best place to configure that to be created ContentFilePool within the conf.xml?

Hmm... I am not certain at the moment I don't think there are any API specific settings in the `conf.xml` file. The closest thing in concept is perhaps the `binary-manager` section, so perhaps a new section that is similar to that? or perhaps conf.xml is not even the right place for this?

reinhapa commented:
> Hmm... I am not certain at the moment I don't think there are any API specific settings in the `conf.xml` file. The closest thing in concept is perhaps the `binary-manager` section, so perhaps a new section that is similar to that? or perhaps conf.xml is not even the right place for this?

At least for the server part I think `conf.xml` seems to be the right place for me. The question is if we would need a new section as it is used within the RPC servlet only...

@dizzzz @wolfgangmm @line-o  any other ideas on this?

reinhapa commented:
I created a separate PR #5195 adding a `ContentFile` pooling mechanism. It still has to be decided, where the configuration actually needs to go.

dizzzz commented on exist-core/src/main/java/org/exist/xmlrpc/RpcAPI.java:
```diff
@@ -926,7 +927,7 @@ String hasUserLock(String path)
     List<String> getGroups() throws EXistException, PermissionDeniedException;
 
-    List<List> getIndexedElements(String collectionName, boolean inclusive)
+    List<List<Object>> getIndexedElements(String collectionName, boolean inclusive)
```
this works? 
<img width="448" alt="image" src="https://github.com/eXist-db/exist/assets/1700062/005b9241-b01e-4b8d-8792-538505226f4c">

dizzzz reviewed (approved):
LGTM....

reinhapa commented:
@adamretter I've integrated the pooling for those ContentFile instances now...

adamretter reviewed (changes requested):
LGTM. Can you address these Sonarcloud issues, then I think it's good to merge:

1. https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=5188&id=eXist-db_exist&open=AYz6Yodr6PRvCQqwhtF0
2. https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=5188&id=eXist-db_exist&open=AYz5OtO43AIjI1_SDloE
3. https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=5188&id=eXist-db_exist&open=AYz8jIkUHDD9adKHbyf3
4. https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=5188&id=eXist-db_exist&open=AY2Dy9Eqk9o22oitb7OP

Also I think these should be improved to be less fragile to time sensitivity (see Collection lock tests for an example), by polling the status or a latch, and a timeout?
1. https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=5188&id=eXist-db_exist&open=AYz5OtOi3AIjI1_SDloC
2. https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=5188&id=eXist-db_exist&open=AYz5OtOj3AIjI1_SDloD

reinhapa commented:
@adamretter I addressed all the changes you mentioned. Could you re-review it and approve in order to merge?

line-o reviewed (approved):
The issues reported by SonarCloud are unrelated to the actual changes in the code.
It was SonarCloud itself that suggested to add the `@Deprecated` annotation and now it complains about missing since and forRemoval. Finding out since which version that method is deprecated is a lot of extra work.  
I accept the changes as is. If you, @reinhapa, decide to add those this is just an extra.

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -480,78 +487,145 @@ public void testQueryWithStylesheet() throws XmlRpcException, MalformedURLExcept
         params.add(query.getBytes(UTF_8));
-        params.add(new HashMap<>());
-        XmlRpcClient xmlrpc = getClient();
+        params.add(Map.of());
```
Perhaps better to use `Collections.emptyMap()` ?

adamretter commented on exist-core/src/main/java/org/exist/util/io/VirtualTempPath.java:
```diff
@@ -52,6 +52,10 @@ public final class VirtualTempPath implements ContentFile {
     @GuardedBy("lock")
     private Path contentFile;
 
+    public VirtualTempPath(TemporaryFileManager tempFileManager) {
```
Param can be `final`.

adamretter commented on exist-core/src/main/java/org/exist/xmlrpc/RpcConnection.java:
```diff
@@ -3057,33 +3035,34 @@ private Map<String, Object> summaryToMap(final long queryTime, @Nullable final S
+                }
+                return result;
+            } else {
+                return List.of();
```
Perhaps `Collections.emptyList()` is better?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -116,58 +127,69 @@ private static String getUri() {
     @After
     public void tearDown() throws XmlRpcException, MalformedURLException {
         XmlRpcClient xmlrpc = getClient();
```
Variable could be `final`

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -116,58 +127,69 @@ private static String getUri() {
     @Test
     public void testStoreAndRetrieve() throws XmlRpcException, IOException {
-        XmlRpcClient xmlrpc = getClient();
+        XmlRpcClient xmlrpc = createCollection(TARGET_COLLECTION);
```
Variable could be `final`

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -116,58 +127,69 @@ private static String getUri() {
-        XmlRpcClient xmlrpc = getClient();
+        XmlRpcClient xmlrpc = createCollection(TARGET_COLLECTION);
+
         List<Object> params = new ArrayList<>();
```
Variable could be `final`

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -116,58 +127,69 @@ private static String getUri() {
         Map table = (Map) xmlrpc.execute("getDocumentData", params);
 
-        try (final UnsynchronizedByteArrayOutputStream os = new UnsynchronizedByteArrayOutputStream()) {
+        try (final ByteArrayOutputStream os = new ByteArrayOutputStream()) {
```
Why replace `UnsynchronizedByteArrayOutputStream` with less performant `ByteArrayOutputStream`?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -116,58 +127,69 @@ private static String getUri() {
+        try (final ByteArrayOutputStream os = new ByteArrayOutputStream()) {
             int offset = (int) table.get("offset");
-            data = (byte[]) table.get("data");
+            byte[] data = (byte[]) table.get("data");
```
Variables could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -179,58 +201,53 @@ public void testStoreAndRetrieve() throws XmlRpcException, IOException {
 
-    private void storeData() throws XmlRpcException, MalformedURLException {
+    private XmlRpcClient createCollection(XmldbURI collection) throws XmlRpcException, MalformedURLException {
         XmlRpcClient xmlrpc = getClient();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -179,58 +201,53 @@ public void testStoreAndRetrieve() throws XmlRpcException, IOException {
 
-        params.clear();
+    private XmlRpcClient storeData() throws XmlRpcException, MalformedURLException {
+        XmlRpcClient xmlrpc = createCollection(TARGET_COLLECTION);
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -179,58 +201,53 @@ public void testStoreAndRetrieve() throws XmlRpcException, IOException {
+    private XmlRpcClient storeData() throws XmlRpcException, MalformedURLException {
+        XmlRpcClient xmlrpc = createCollection(TARGET_COLLECTION);
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -179,58 +201,53 @@ public void testStoreAndRetrieve() throws XmlRpcException, IOException {
 
-        params.clear();
-        final String generatedXml = generateXml((int)(MAX_DOWNLOAD_CHUNK_SIZE * 1.5));
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -249,30 +266,25 @@ public void getDocumentDataChunked_nextChunk() throws IOException, XmlRpcExcepti
 
-        params.clear();
-        final String generatedXml = generateXml((int)(MAX_DOWNLOAD_CHUNK_SIZE * 1.75));
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -356,46 +368,36 @@ public void uploadCompressedAndDownload() throws IOException, XmlRpcException {
-        XmlRpcClient xmlrpc = getClient();
+        final XmlRpcClient xmlrpc = storeData();
+
         List params = new ArrayList(1);
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -356,46 +368,36 @@ public void uploadCompressedAndDownload() throws IOException, XmlRpcException {
-        XmlRpcClient xmlrpc = getClient();
+        final XmlRpcClient xmlrpc = storeData();
+
         List params = new ArrayList(1);
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -356,46 +368,36 @@ public void uploadCompressedAndDownload() throws IOException, XmlRpcException {
-        storeData();
+        final XmlRpcClient xmlrpc = storeData();
+
         Map<String, String> options = new HashMap<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -407,50 +409,56 @@ public void testRetrieveDoc() throws XmlRpcException, MalformedURLException {
-        storeData();
+        final XmlRpcClient xmlrpc = storeData();
+
         List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -407,50 +409,56 @@ public void testRetrieveDoc() throws XmlRpcException, MalformedURLException {
+    public void testQuerySummary() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -407,50 +409,56 @@ public void testRetrieveDoc() throws XmlRpcException, MalformedURLException {
-        storeData();
+        final XmlRpcClient xmlrpc = storeData();
+
         Map<String, String> options = new HashMap<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -480,78 +487,145 @@ public void testQueryWithStylesheet() throws XmlRpcException, MalformedURLExcept
-        storeData();
+        final XmlRpcClient xmlrpc = storeData();
+
         List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -480,78 +487,145 @@ public void testQueryWithStylesheet() throws XmlRpcException, MalformedURLExcept
+    public void testAccount() throws MalformedURLException, XmlRpcException {
+        final String user = "rudi";
+        final String passwd = "pass";
         String simpleMd5 = MessageDigester.md5(passwd, true);
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -480,78 +487,145 @@ public void testQueryWithStylesheet() throws XmlRpcException, MalformedURLExcept
+
+        assertThat(xmlrpc.execute("setUserPrimaryGroup", List.of(user, "guest"))).isEqualTo(TRUE);
 
         XmlRpcClientConfigImpl config = (XmlRpcClientConfigImpl) xmlrpc.getClientConfig();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -480,78 +487,145 @@ public void testQueryWithStylesheet() throws XmlRpcException, MalformedURLExcept
-        storeData();
+        final XmlRpcClient xmlrpc = storeData();
+
         List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -480,78 +487,145 @@ public void testQueryWithStylesheet() throws XmlRpcException, MalformedURLExcept
-        storeData();
+        final XmlRpcClient xmlrpc = storeData();
+
         List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -609,7 +673,7 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
-        assertTrue("added collection not found", foundMatch);
+        assertThat(foundMatch).withFailMessage("added collection not found").isTrue();
 
         Map<String, String> options = new HashMap<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+
+    @Test
+    public void testGetVersion() throws XmlRpcException, MalformedURLException {
+        XmlRpcClient xmlrpc = getClient();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    @Test
+    public void testGetVersion() throws XmlRpcException, MalformedURLException {
+        XmlRpcClient xmlrpc = getClient();
+        assertThat(xmlrpc.execute("getVersion", List.of())).isEqualTo(Version.getVersion());
```
Perhaps  better to use `Collections.emptyList()`?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testConfigureCollection() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = createCollection(TARGET_COLLECTION);
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testCreateId() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = createCollection(TARGET_COLLECTION);
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testCreateResourceId() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = createCollection(TARGET_COLLECTION);
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testGetCollectionDesc() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variables could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testExistsAndCanOpenCollection() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variables could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testDescribeResource() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variables could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testGetContentDigest() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variables could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+
+        assertThat((Object[]) xmlrpc.execute("getDocumentListing", List.of())).isNotEmpty();
+
+        List<Object> params = new ArrayList<>();
```
Variables could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testGetCollectionListing() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variables could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testGetResourceCount() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variables could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testChown() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testGetBinaryResource() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testListDocumentPermissions() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testListCollectionPermissions() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testGetSubCollectionCreationTime() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testGetSubCollectionPermissions() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testQueryPT() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testQueryP() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testPrintDiagnostics() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testGetCreationDate() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter commented on exist-core/src/test/java/org/exist/xmlrpc/XmlRpcTest.java:
```diff
@@ -622,7 +686,331 @@ public void testCollectionWithAccentsAndSpaces() throws XmlRpcException, Malform
+    public void testGetSubResourcePermissions() throws XmlRpcException, MalformedURLException {
+        final XmlRpcClient xmlrpc = storeData();
+
+        List<Object> params = new ArrayList<>();
```
Variable could be final?

adamretter reviewed (commented):
Great piece of work thanks @reinhapa :-)
Just 3 observations:

1. Some variables/params could be marked `final` please
2. Empty collections are maybe better created with Collections.emptyXXX to make intention clear.
3. Not sure why the change from `UnsynchronizedByteArrayOutputStream` to the less performant `ByteArrayOutputStream`. I spent time previously chasing out ByteArrayOutput/InputStream from the code base.

reinhapa commented:
Hi @adamretter 

> * Some variables/params could be marked `final` please

Most places where already that way before my change but should now be changed.

> * Empty collections are maybe better created with Collections.emptyXXX to make intention clear.

I've changed those accordingly
 
> * Not sure why the change from `UnsynchronizedByteArrayOutputStream` to the less performant `ByteArrayOutputStream`. I spent time previously chasing out ByteArrayOutput/InputStream from the code base.

That was done because of the deprecation warning... I reverted that change, we might need to fix those warnings by other means...
